### PR TITLE
[Multi Hub] Interact with calendar, mail, profile, geolocation from personal tabs app

### PR DESCRIPTION
**Pull request by Harikrishnan-MSFT**
… personal tabs

### CODE REVIEW

**Comment by Harikrishnan-MSFT** [comment]
Build Pipeline -
https://github.com/OfficeDev/Microsoft-Teams-Samples/actions/runs/5455275342
image

File: samples/tab-multi-hub/ts/README.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+      * **Admin consent description**: Allows Teams to call the app’s web APIs as the current user.
+      * **User consent title**: Teams can access the user profile and make requests on the user's behalf.
+      * **User consent description:** Enable Teams to call this app’s APIs with the same rights as the user.
+  9. Ensure that **State** is set to **Enabled**
+  10. Select **Add scope**
+      * The domain part of the **Scope name** displayed just below the text field should automatically match the **Application ID** URI set in the previous step, with `/access_as_user` appended to the end:
+          * `api://[ngrokDomain].ngrok-free.app/00000000-0000-0000-0000-000000000000/access_as_user.
+  11. In the **Authorized client applications** section, identify the applications that you want to authorize for your app’s web application. Each of the following IDs needs to be entered:
+         * `4765445b-32c6-49b0-83e6-1d93765276ca` (Office web)
+         * `0ec893e0-5785-4de6-99da-4ed124e5296c` (Office desktop)
```
Comment (ChetanSharma-msft) [inline samples/tab-multi-hub/ts/README.md:65]: MObile client id's are missing.

File: samples/tab-multi-hub/ts/README.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+**Click the button to get your current geolocation:**
+
+![M6](Images/mobile/M9.png)
+
+## Further Reading.
+  
+- [TeamsJS capability support across Microsoft 365](https://learn.microsoft.com/en-us/microsoftteams/platform/m365-apps/teamsjs-support-m365)
+- [Extend Teams apps across Microsoft 365](https://learn.microsoft.com/microsoftteams/platform/m365-apps/overview)
+
+<img src="https://pnptelemetry.azurewebsites.net/microsoft-teams-samples/samples/tab-calendar-ts" />
```
Comment (ChetanSharma-msft) [inline samples/tab-multi-hub/ts/README.md:254]: This is still wrong:
/tab-calendar-ts"

File: samples/tab-multi-hub/ts/api-server/server.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+require('dotenv').config({ path: ENV_FILE });
+
+const clientId = process.env.CLIENT_ID;
+const clientSecret = process.env.CLIENT_SECRET;
+const graphScopes = ['https://graph.microsoft.com/' + process.env.GRAPH_SCOPES];
+let handleQueryError = function (err) {
+    console.log("handleQueryError called: ", err);
+    return new Response(JSON.stringify({
+        code: 400,
+        message: 'Stupid network Error'
```
Comment (ChetanSharma-msft) [inline samples/tab-multi-hub/ts/api-server/server.js:17]: Please update it properly

File: samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import { app, geoLocation } from "@microsoft/teams-js";
+
+function GeolocationTab() {
+  const [isLocationVisible, setIsLocationVisible] = useState(false);
+  const [platformIsSupported, setplatformIsSupported] = useState(false);
+  const [currentLocation, setCurrentLocation] = useState <geoLocation.Location>();
+  let hasConsent : Boolean = false
+  useEffect(() => {
+    app.initialize()
+  })
```
Comment (ChetanSharma-msft) [inline samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx:14]: Spacing

File: samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  const getGeolocation = async () => {
+    try
+    {
+      if (geoLocation.isSupported()) {
+        const hasPerms = await geoLocation.hasPermission();
+        if(!hasPerms)
```
Comment (ChetanSharma-msft) [inline samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx:20]: Spacing

File: samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      }
+      else 
+      {
+        setplatformIsSupported(true);
+      }
+    } catch (e) {
+      console.log(`GeoLocation error: ${e}`);
+    }
+  }
+  return (
```
Comment (ChetanSharma-msft) [inline samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx:42]: Spacing

File: samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      if (geoLocation.isSupported()) {
+        const hasPerms = await geoLocation.hasPermission();
+        if(!hasPerms)
+        {
+           hasConsent =  await geoLocation.requestPermission();
+        }
+        else{
+          hasConsent = true
+        }
+        if(hasConsent)
```
Comment (ChetanSharma-msft) [inline samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx:27]: Spacing

File: samples/tab-multi-hub/ts/src/components/MailTab.tsx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+  // Set value for To Recepients.
+  const functionRecepients = (e: any) => {
+    let email = e.target.value;
+    let flag: boolean = true;
+    const toEmailList: [] = email.split(";");
+
+    // checking valid email from the list 
+    toEmailList.map(e => {
+      if (validator.isEmail(e)) {
```
Comment (ChetanSharma-msft) [inline samples/tab-multi-hub/ts/src/components/MailTab.tsx:32]: Spacing

File: samples/tab-multi-hub/ts/src/components/Tab.tsx [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+          console.error("SSO failed: ", error);
+        });
+    });
+  }
+
+  // Exchange the SSO access token for a Graph access token
+  // Learn more: https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow
+  exchangeClientTokenForServerToken = async (token: string) => {
+    let serverURL = `${process.env.REACT_APP_BASE_URL}/getGraphAccessToken?ssoToken=${token}&upn=${this.state.context?.user?.userPrincipalName}`;
+    let response = await fetch(serverURL).catch(this.unhandledFetchError); //This calls getGraphAccessToken route in /api-server/app.js
```
Comment (ChetanSharma-msft) [inline samples/tab-multi-hub/ts/src/components/Tab.tsx:92]: Spacing

File: samples/tab-multi-hub/ts/src/components/Tab.tsx [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+    var OpenCalendarItemParams = {
+      itemId: itemIds
+    };
+    if (microsoftTeams.calendar.isSupported()) {
+      microsoftTeams.calendar.openCalendarItem(OpenCalendarItemParams);
+    } else {
+      this.setState({ platformIsSupported: true });
+    }
+  }
+  composeMeeting() {
```
Comment (ChetanSharma-msft) [inline samples/tab-multi-hub/ts/src/components/Tab.tsx:226]: Spacing

File: samples/tab-multi-hub/ts/src/components/TabCalendarMain.tsx [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+import React from 'react';
+import './App.css';
+
+// This component is used to display the required privacy statement which can be found in a link in the about tab.
+class TabCalendar extends React.Component {
+    render() {
+      return (
+        <div>
+          <h2>Welcome to interacting with the user's calendar, mail, profile from personal tabs app in Outlook.</h2>
```
Comment (ChetanSharma-msft) [inline samples/tab-multi-hub/ts/src/components/TabCalendarMain.tsx:13]: Please update it.

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Added review comments.

File: samples/tab-multi-hub/ts/README.md (hunk 1 above)
Comment (Harikrishnan-MSFT) [reply samples/tab-multi-hub/ts/README.md:65]: Fixed

File: samples/tab-multi-hub/ts/README.md (hunk 2 above)
Comment (Harikrishnan-MSFT) [reply samples/tab-multi-hub/ts/README.md:254]: Fixed

File: samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx (hunk 4 above)
Comment (Harikrishnan-MSFT) [reply samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx:14]: Fixed

File: samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx (hunk 5 above)
Comment (Harikrishnan-MSFT) [reply samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx:20]: Fixed

File: samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx (hunk 6 above)
Comment (Harikrishnan-MSFT) [reply samples/tab-multi-hub/ts/src/components/GeolocationTab.tsx:42]: Fixed

File: samples/tab-multi-hub/ts/src/components/MailTab.tsx (hunk 8 above)
Comment (Harikrishnan-MSFT) [reply samples/tab-multi-hub/ts/src/components/MailTab.tsx:32]: Fixed

File: samples/tab-multi-hub/ts/src/components/Tab.tsx (hunk 9 above)
Comment (Harikrishnan-MSFT) [reply samples/tab-multi-hub/ts/src/components/Tab.tsx:92]: Fixed

File: samples/tab-multi-hub/ts/src/components/Tab.tsx (hunk 10 above)
Comment (Harikrishnan-MSFT) [reply samples/tab-multi-hub/ts/src/components/Tab.tsx:226]: Fixed

File: samples/tab-multi-hub/ts/src/components/TabCalendarMain.tsx (hunk 11 above)
Comment (Harikrishnan-MSFT) [reply samples/tab-multi-hub/ts/src/components/TabCalendarMain.tsx:13]: Fixed

**Review by ChetanSharma-msft — APPROVED** [review]
Changes looks fine, approving it.

**Comment by Harikrishnan-MSFT** [comment]
Build Pipeline -
https://github.com/OfficeDev/Microsoft-Teams-Samples/actions/runs/5456061201
image

**Review by Wajeed-msft — APPROVED** [review]

**Pull request merged**